cmd/artifact/main.go

```diff
@@ -0,0 +1,341 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	tcclient "github.com/taskcluster/taskcluster-client-go"
+	"github.com/taskcluster/taskcluster-client-go/queue"
+	artifact "github.com/taskcluster/taskcluster-lib-artifact-go"
+	"github.com/urfave/cli"
+)
+
+const (
+	ErrInternal = 1
+	ErrBadUsage = 2
+	ErrCorrupt  = 13
+)
+
+// TODO implement an in memory 'file'
+// TODO implement 'redirect' and 'error' artifact types?
+
+func main() {
+	err := _main(os.Args)
+	if err == nil {
+		os.Exit(0)
+	}
+
+	if ecErr, ok := err.(cli.ExitCoder); ok {
+		os.Exit(ecErr.ExitCode())
+	}
+
+	os.Exit(ErrInternal)
+
+}
+
+func _main(args []string) error {
+	// We're going to take care of exiting ourselves
+	cli.OsExiter = func(c int) {}
+
+	app := cli.NewApp()
+
+	app.Name = "artifact"
+	app.Version = "0.0.1"
+	app.Usage = "interact with taskcluster artifacts"
+
+	app.OnUsageError = func(c *cli.Context, err error, isSubcommand bool) error {
+		if isSubcommand {
+			return err
+		}
+
+		fmt.Fprintf(c.App.Writer, "WRONG: %#v\n", err)
+		return nil
+	}
+
+	app.Action = func(c *cli.Context) error {
+		cli.ShowAppHelp(c)
+		if c.NArg() == 0 {
+			return cli.NewExitError("Must specify command", ErrBadUsage)
+		}
+		return cli.NewExitError(fmt.Sprintf("%s is not a command", c.Args().Get(0)), ErrBadUsage)
+	}
+
+	app.OnUsageError = func(context *cli.Context, err error, isSubcommand bool) error {
+		return cli.NewExitError(err, ErrBadUsage)
+	}
+
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:   "client-id",
+			EnvVar: "TASKCLUSTER_CLIENT_ID",
+			Usage:  "set client id to `CLIENT_ID`",
+		},
+		cli.StringFlag{
+			Name:   "access-token",
+			EnvVar: "TASKCLUSTER_ACCESS_TOKEN",
+			Usage:  "set access token to `ACCESS_TOKEN`",
+		},
+		cli.StringFlag{
+			Name:   "certificate",
+			EnvVar: "TASKCLUSTER_CERTIFICATE",
+			Usage:  "set certificate to `CERTIFICATE`",
+		},
+		cli.StringFlag{
+			Name: "base-url",
+			//EnvVar: "QUEUE_BASE_URL",
+			Usage: "set queue's `BASE_URL`",
+			Value: queue.DefaultBaseURL,
+		},
+		cli.IntFlag{
+			Name:   "chunk-size",
+			Usage:  "set the I/O chunk size to `CHUNK_SIZE` KB",
```

Comment:
Really this should say [KiB](https://physics.nist.gov/cuu/Units/binary.html) according to SI units.

An alternative would be to use `cli.Int64Flag` and specify all content sizes in bytes. This would help with consistency across the command line parameters. It also is in keeping with e.g. `Content-Length` header being in bytes. Another alternative might be to allow the parameter to specify units (e.g. `--chunk-size 5 MiB`).

Reply:
FWIW I had a play, and this looks like a useful library for this purpose:

```go
pmoore@Petes-iMac:~/go/src/github.com/taskcluster/testparse $ cat main.go 
package main

import (
	"fmt"
	"log"
	"strings"

	"github.com/alecthomas/units"
)

func main() {
	const size = "24.3 MiB"
	bytes, err := units.ParseBase2Bytes(strings.Replace(size, " ", "", -1))
	if err != nil {
		log.Fatalf("Woooah: %v", err)
	}
	fmt.Printf("%v == %#v bytes\n", size, bytes)
}
pmoore@Petes-iMac:~/go/src/github.com/taskcluster/testparse $ go run main.go 
24.3 MiB == 25480396 bytes
pmoore@Petes-iMac:~/go/src/github.com/taskcluster/testparse $ 
```

Reply:
There's some debate about whether KiB or KB, as opposed to kB is the best option.  I personally find KiB ugly, so tend to use KB, plus the only people who really care about 1000 byte increments are those short changing us on physical storage :(

Regarding the chunk size, it would be in line with Content-Length, but based on benchmarks, anything lower than about 32KB is so slow it'd be useless, which means that we'd be having to do things like `--chunk-size 65536` instead of `--chunk-size 64`.

Similarly, for multi-part, the S3 enforced sizes are 5MB-5GB, which means that we'd always be passing 5242880-5368709120 instead of 5-5000.

Both of these options are not intended to be used commonly, but I think the pragmatism of using the KB and MB is worthwhile.  Alternately, do you know a good go library which does parsing of "1 KB" into a number 1024?  That'd be a nicer solution all around!

Reply:
See code sample above - I really like github.com/alecthomas/units. Funnily enough he is also the author of gometalinter and also the python voluptuous library.